### PR TITLE
Support POST for /awkward/buffers/

### LINKED ISSUE #574: Reading Awkward Arrays requires URLs that are too long
I don't remember where I mentioned this[^1], but one thing that I'm using this for is to collect metadata for a set of ROOT files, which can then be accessed more quickly (like [kerchunk](https://fsspec.github.io/kerchunk/), but ROOT metadata is big enough that it can't be JSON).

I started this project (https://github.com/jpivarski/tiled-uproot), using Awkward Arrays in Tiled for the metadata. For each PB of ROOT data, I estimate about 10 TB of metadata will be needed in the database. (User derived quantities will be smaller, and it will be nice to be able to access the original ROOT data and derived quantities with the same tools.)

I ran this on 3 typical files, resulting in an Awkward Array whose Form JSON is 786 kB, has 5639 buffers containing a total of 42 MB. With more files, the size of the Form and number of buffers will not increase, but the total amount of data in the buffers will increase.

I had no trouble writing this Awkward Array in Tiled, and I could read back small slices of it. However, some queries, the ones addressing the larger parts of the array, raised this error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/tiled/client/awkward.py", line 55, in __getitem__
    return self.read(slice=slice)
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/tiled/client/awkward.py", line 39, in read
    self.context.http_client.get(
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_client.py", line 1041, in get
    return self.request(
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_client.py", line 801, in request
    request = self.build_request(
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_client.py", line 358, in build_request
    return Request(
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_models.py", line 330, in __init__
    self.url = self.url.copy_merge_params(params=params)
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_urls.py", line 360, in copy_merge_params
    return self.copy_with(params=self.params.merge(params))
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_urls.py", line 348, in copy_with
    return URL(self, **kwargs)
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_urls.py", line 115, in __init__
    self._uri_reference = url._uri_reference.copy_with(**kwargs)
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_urlparse.py", line 134, in copy_with
    return urlparse("", **defaults)
  File "/home/jpivarski/mambaforge/envs/tiled/lib/python3.10/site-packages/httpx/_urlparse.py", line 200, in urlparse
    raise InvalidURL(f"URL component '{key}' too long")
```

Given this scale of Form JSON, maybe the URL-based query can be replaced by POST data?

[^1]: Finally found it: it was a private email with subject line "Using Tiled as a kerchunk-like database for ROOT files". I copied it here: https://github.com/jpivarski/tiled-uproot/discussions/1.

### PR DESCRIPTION
**Pull request by danielballan**
Closes #574 

At your convenience, would you take this for a spin, @jpivarski, and confirm it works on your large AwkwardArrays?

### CODE REVIEW

**Comment by jpivarski** [comment]
This PR fixes the

```
httpx.InvalidURL: URL component 'query' too long
```

error! I just tried it in `main` and in this PR branch, and it toggles the error.

**Comment by danielballan** [comment]
Great. I have also added a unit test here that reproduces the issue.

File: tiled/_tests/test_awkward.py [hunk 1]
```diff
@@ -116,3 +116,14 @@ def test_export_parquet(tmpdir):
...
+
+def test_long_url(tmpdir):
+    # https://github.com/bluesky/tiled/pull/577
+    catalog = in_memory(writable_storage=tmpdir)
+    app = build_app(catalog)
+    with Context.from_app(app) as context:
+        client = from_context(context)
+        array = awkward.Array([{f"key{i:05}": i for i in range(4000)}])
+        aac = client.write_awkward(array, key="test")
+        aac[0]
```
Comment (padraic-shafer) [inline tiled/_tests/test_awkward.py:129]: This test induces the error in the main branch and passes with the accompanying codebase changes in the PR.  However, the reason for failing or passing the test is abstruse--it's obscured by what's happening under the hood in `AwkwardClient.read()`.

I have some ideas how to clarify this, but I've "timed out" for the day.  I'd like to push a suggestion tomorrow.

File: tiled/client/awkward.py [hunk 2]
```diff
@@ -34,12 +34,12 @@ def read(self, slice=...):
...
         # The order is not important, but sort so that the request is deterministic.
-        params = {"form_key": sorted(list(form_keys_touched))}
+        form_keys = sorted(list(form_keys_touched))
         content = handle_error(
-            self.context.http_client.get(
+            self.context.http_client.post(
                 self.item["links"]["buffers"],
                 headers={"Accept": "application/zip"},
-                params=params,
+                json=form_keys,
```
Comment (padraic-shafer) [inline tiled/client/awkward.py:42]: These changes look good.

File: tiled/server/router.py [hunk 3]
```diff
@@ -619,7 +619,69 @@ async def awkward_buffers(
...
+        filename=filename,
+        serialization_registry=serialization_registry,
+        settings=settings,
+    )
+
+
+async def _awkward_buffers(
+    request: Request,
+    entry=SecureEntry(scopes=["read:data"]),
+    form_key: Optional[List[str]] = Query(None, min_length=1),
```
Comment (padraic-shafer) [inline tiled/server/router.py:679]: I'm not sure whether the default value here (`Query(...)`) has any meaning for parameters of a function that is not decorated with an `APIRoute`.

File: tiled/server/router.py [hunk 4]
```diff
@@ -619,7 +619,69 @@ async def awkward_buffers(
...
+    format: Optional[str] = None,
+    filename: Optional[str] = None,
+    serialization_registry=Depends(get_serialization_registry),
+    settings: BaseSettings = Depends(get_settings),
```
Comment (padraic-shafer) [inline tiled/server/router.py:683]: I'm not sure whether the default values here (`Depends(...)`) have any meaning for parameters of a function that is not decorated with an `APIRoute`.

**Review by padraic-shafer — COMMENTED** [review]
* I have some suggestions for improving the test.
* I had intended to test the default parameters for `tiled.server.router._awkward_buffers()` but did not yet get the chance.

Will add that feedback tomorrow.

File: tiled/server/router.py (hunk 3 above)
Comment (danielballan) [reply tiled/server/router.py]: Fair point; this is just noise. I've stripped off all the defaults in the latest commit.

File: tiled/_tests/test_awkward.py [hunk 5]
```diff
@@ -3,116 +3,172 @@
...
+        ("Safari", 80_000),
+        # ("Internet Explorer", 2_083),
+    )
+)
+def test_long_url(spy_client, spy, url_length_limit, browser, num_keys=7_000):
+    # https://github.com/bluesky/tiled/pull/577
+    array = awkward.Array([{f"key{i:05}": i for i in range(num_keys)}])
+    aac = spy_client.write_awkward(array, key="test")
+    aac[0]
+
```
Comment (padraic-shafer) [inline tiled/_tests/test_awkward.py:168]: I added a check on the form_keys length here.  It's a bit invasive on the implementation, but this is one way to show what is being tested here.

Comment (padraic-shafer) [reply tiled/_tests/test_awkward.py:168]: To make this change simpler, I refactored the tests to use a client fixture.

Feel free to pop this commit.  I added it just to show my thinking.

**Review by padraic-shafer — COMMENTED** [review]
I added a check on the form_keys length here. It's a bit invasive on the implementation, but this is one way to show what is being tested here.

To make this change simpler, I refactored the tests to use a client fixture.

Feel free to pop this commit. I added it just to show my thinking.

**Comment by danielballan** [comment]
Thanks @padraic-shafer. I have pushed my counter-proposal:

* Inspecting the Request is good. Let's use httpx's own mechanism (wrapped by Tiled in a context manage, `record_history`, for convenience) instead of Mock.
* I'm mindful of keeping the test suite fast, and I think picking one URL length limit is enough to make our point. That's not a strongly held view...I did this in a separate commit that we can revert if you disagree.

**Comment by danielballan** [comment]
Oh, and I should add, the fixtures are a great improvement. We do that elsewhere...not sure why I got lazy in this case.

**Review by padraic-shafer — APPROVED** [review]
This is ready to go.
* Long URL issue #574 for Awkward arrays is resolved
* The associated test demonstrates how the issue was resolved (i.e., why the POST route exists for fetching data)

**Pull request merged**